src/main/scala/mesosphere/marathon/core/storage/StorageModule.scala

```diff
@@ -22,68 +19,96 @@ import scala.concurrent.ExecutionContext
   * Provides the repositories for all persistable entities.
   */
 trait StorageModule {
-  def appRepository: AppRepository
+  def appRepository: ReadOnlyAppRepository
   def taskRepository: TaskRepository
   def deploymentRepository: DeploymentRepository
   def taskFailureRepository: TaskFailureRepository
+  def groupRepository: GroupRepository
+  def migration: Migration
 }
 
 object StorageModule {
-  def apply(conf: MarathonConf, metrics: Metrics, mat: Materializer, ctx: ExecutionContext,
-    scheduler: Scheduler, actorRefFactory: ActorRefFactory): StorageModule =
-    apply(StorageConfig(conf))(metrics, mat, ctx, scheduler, actorRefFactory)
+  def apply(conf: MarathonConf)(implicit metrics: Metrics, mat: Materializer, ctx: ExecutionContext,
+    scheduler: Scheduler, actorRefFactory: ActorRefFactory): StorageModule = {
+    val currentConfig = StorageConfig(conf)
+    val legacyConfig = conf.internalStoreBackend() match {
+      case TwitterZk.StoreName => Some(TwitterZk(cache = true, conf))
```

Comment:
For migration purposes, I figured having caching on would be very helpful. The legacyConfig (i realize its confusing here - this area needs a lot more cleanup still) is used solely to move from pre-1.3 -> 1.3 - and the migration code closes the underlying store when its done -- needs a lot more documentation.